src/java.base/share/classes/jdk/internal/util/ClassFileDumper.java

```diff
@@ -0,0 +1,215 @@
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Helper class to log normal and hidden classes defined via Lookup::defineClass
+ * and Lookup::defineHiddenClass API
+ *
+ * @implNote
+ * <p> Because this class is called by MethodHandleStatics, LambdaForms generation
+ * and LambdaMetafactory, make use of lambda lead to recursive calls cause stack overflow.
+ */
+public final class ClassFileDumper {
+    private static final ConcurrentHashMap<String, ClassFileDumper> DUMPER_MAP
+            = new ConcurrentHashMap<>();
+
+    /**
+     * Returns a ClassFileDumper instance for the given key.  To enable
+     * dumping of the generated classes, set the system property via
+     * -D<key>=<path>.
+     *
+     * The system property is read only once when it is the first time
+     * the dumper instance for the given key is created.
+     *
+     * If not enabled, this method returns ClassFileDumper with null
+     * dump path.
+     */
+    public static ClassFileDumper getInstance(String key) {
+        Objects.requireNonNull(key);
+
+        var dumper = DUMPER_MAP.get(key);
+        if (dumper == null) {
+            String path = GetPropertyAction.privilegedGetProperty(key);
+            Path dir;
+            if (path == null || path.trim().isEmpty()) {
+                dir = null;
+            } else {
+                dir = validateDumpDir(Path.of(path.trim()));
+            }
+            var newDumper = new ClassFileDumper(key, dir);
+            var v = DUMPER_MAP.putIfAbsent(key, newDumper);
+            dumper = v != null ? v : newDumper;
+        }
+        return dumper;
+    }
+
+    /**
+     * Returns a ClassFileDumper instance for the given key with a given
+     * dump path. To enable dumping of the generated classes
+     *     -D<key> or -D<key>=true
+     *
+     * The system property is read only once when it is the first time
+     * the dumper instance for the given key is created.
+     *
+     * If not enabled, this method returns ClassFileDumper with null
+     * dump path.
+     */
+    public static ClassFileDumper getInstance(String key, Path path) {
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(path);
+
+        var dumper = DUMPER_MAP.get(key);
+        if (dumper == null) {
+            String value = GetPropertyAction.privilegedGetProperty(key);
+            boolean enabled = value != null && value.isEmpty()
+                                    ? true : Boolean.parseBoolean(value);
+            Path dir = enabled ? validateDumpDir(path) : null;
+            var newDumper = new ClassFileDumper(key, dir);
+            var v = DUMPER_MAP.putIfAbsent(key, newDumper);
+            dumper = v != null ? v : newDumper;
+        }
+
+        if (dumper.isEnabled() && !path.equals(dumper.dumpPath())) {
+            throw new IllegalArgumentException("mismatched dump path for " + key);
+        }
```

Comment:
I don't see how this exception case could ever occur, given that `dumper.dumpPath` is directly coming from `dir`, and `validateDumpDir` doesn't return a modified path either.

Can this check be removed?

Reply:
This is to check if two callers get the `ClassFileDumper` of the same key but different path.

```
  ClassFileDumper.getInstance("jdk.foo.dump", "DUMP_DIR");
  ClassFileDumper.getInstance("jdk.foo.dump", "DUMP_NEW_DIR");  <--- IAE
```

Reply:
Ok, I see.